File: golang/calcpass/parallel_bcrypt/parallel_bcrypt.go
/**Run bcrypt password hash in parallel threads and combine the results.
Four threads is a good choice.  Since most modern computers and
 mobile devices have 4 core processors the hash completes in about
 the same amount of time as a single bcrypt, however it makes
 cracking attempts 4X more costly.
*/
package parallel_bcrypt

import (
	"github.com/cruxic/calcpass/golang/calcpass/util"
	"golang.org/x/crypto/bcrypt"
	"crypto/sha256"
	"errors"
	"strconv"
)

//Salt must be exactly this number of bytes
const BcryptSaltLen = 16

//The number of bytes returned by Hash()
const OutputSize = 32

type thread_result struct {
	threadIndex int	
	bcryptHash []byte
	err error
}

/**Hash given password with N threads (1-32).  There is no length limit on the password
because it is first hashed with sha256 before it goes into bcrypt.
The output hash is always 32 bytes.
*/
func Hash(nThreads int, plaintextPassword []byte, salt []byte, cost int) ([]byte, error) {
	if len(plaintextPassword) == 0 {
		return nil, errors.New("empty password or salt");
	}

	if len(salt) != BcryptSaltLen {
		return nil, errors.New("bcrypt salt must be exactly " + strconv.Itoa(BcryptSaltLen) + " bytes")
	}

	if nThreads < 0 || nThreads > 32 {
		return nil, errors.New("nThreads out of range")
	}


	//
	// Spawn the threads
	
	resultChan := make(chan *thread_result, nThreads)
	for i := 0; i < nThreads; i++ {
		go bcryptThread(plaintextPassword, salt, cost, i, resultChan)
	}

	//
	// Wait for threads to finish
	
	bcryptHashes := make([][]byte, nThreads)
	for _ = range bcryptHashes {
		res := <-resultChan
		if res.err != nil {
			//unlikely (bad cost parameter?)
			return nil, res.err
		}
		
		//sanity
		if bcryptHashes[res.threadIndex] != nil {
			panic("duplicate thread index")
		}
		
		bcryptHashes[res.threadIndex] = res.bcryptHash
	}
	
	//Combine all the base64 bcrypt hashes with sha256
	var sha = sha256.New()
	for i := range bcryptHashes {
		sha.Write(bcryptHashes[i])
		util.Erase(bcryptHashes[i])
	}
	
	finalHash := sha.Sum(nil)
	sha.Reset()
	
	return finalHash, nil
}

//Same as hex.EncodeToString() but it returns a []byte instead of string so
// that we can later erase it.
func bytesToHex(data []byte) []byte {
	//ASCII 0-9 a-f
	chars := [16]byte{0x30,0x31,0x32,0x33,0x34,0x35,0x36,0x37,0x38,0x39,0x61,0x62,0x63,0x64,0x65,0x66}

	res := make([]byte, len(data) * 2)
	j := 0
	for _, b := range data {
		res[j] = chars[b >> 4]
		j++
		res[j] = chars[b & 0x0f]
		j++
	}

	return res
}


func bcryptThread(plaintextPassword, salt []byte, cost, threadIndex int, result chan *thread_result) {

	//Derive a distinct password for this thread to work on:
	//  eg: sha256(plaintextPassword + 0x01)
	var sha = sha256.New()
	sha.Write(plaintextPassword)
	sha.Write([]byte{byte(threadIndex + 1)})
	threadPassword := sha.Sum(nil)
	sha.Reset()

	//Some bcrypt implementations are broken (eg PHP) because they truncate
	// the password at the first null byte!  Therefore I'll pass 64 hex characters.
	//(bcrypt can handle up to 72 bytes)
	hexPass := bytesToHex(threadPassword)

	var tr thread_result
	tr.threadIndex = threadIndex
	tr.bcryptHash, tr.err = bcrypt.GenerateFromPasswordAndSalt(hexPass, salt, cost)

	//remove the salt and cost prefix (first 29 chars)
	if tr.err == nil {
		util.Erase(tr.bcryptHash[0:29])
		tr.bcryptHash = tr.bcryptHash[29:]
	}

	util.Erase(threadPassword)
	util.Erase(hexPass)
	
	result <- &tr
}

File: java/src/com/calcpass/parallel_bcrypt/ParallelBcrypt.java
package com.calcpass.parallel_bcrypt;

import com.calcpass.util.Util;
import org.mindrot.jbcrypt.BCrypt;
import java.util.Arrays;
import java.io.UnsupportedEncodingException;

public class ParallelBcrypt {
	public static final int BCRYPT_SALT_SIZE = 16;

	
	static String standard_bcrypt(String password, final byte[] salt, int cost) {
		//A slightly hacky way to tell jBcrypt which salt to use.
		// I considered using Bcrypt.crypt_raw() but it would take more work.
		java.security.SecureRandom saltSource = new java.security.SecureRandom() {
			public void nextBytes(byte[] buffer) {
				if (buffer.length != BCRYPT_SALT_SIZE || salt.length != BCRYPT_SALT_SIZE)
					throw new IllegalArgumentException("wrong salt size!");

				System.arraycopy(salt, 0, buffer, 0, BCRYPT_SALT_SIZE);
			}
		};

		//Encode the cost and salt as bcrypt-base64
		String prefixAndSalt = BCrypt.gensalt(cost, saltSource);

		return BCrypt.hashpw(password, prefixAndSalt);
	}

	

	public static byte[] Hash(int nThreads, byte[] plainTextPassword, byte[] salt, int cost) {
		if (plainTextPassword.length == 0)
			throw new IllegalArgumentException("Empty password");

		if (salt.length != BCRYPT_SALT_SIZE)
			throw new IllegalArgumentException("Wrong bcrypt salt length");

		if (nThreads < 1 || nThreads > 32)
			throw new IllegalArgumentException("nThreads out of range");

		ParallelBcryptThread[] threads = new ParallelBcryptThread[nThreads];

		//Spawn each thread
		for (int i = 0; i < nThreads; i++) {
			threads[i] = new ParallelBcryptThread(plainTextPassword, i, salt, cost);
			threads[i].start();			
		}

		//Wait each thread to finish and concat eash result
		StringBuilder sb = new StringBuilder(nThreads * 128);
		for (int i = 0; i < nThreads; i++) {
			try {
				threads[i].join();
			}
			catch (InterruptedException ie) {
				throw new RuntimeException(ie);
			}

			if (threads[i].failReason != null)
				throw threads[i].failReason;

			sb.append(threads[i].result);
		}

		byte[] allHashes;
		try {
			allHashes = sb.toString().getBytes("UTF-8");
		}
		catch (UnsupportedEncodingException uee) {
			throw new RuntimeException(uee);
		}

		return Util.sha256(allHashes);
	}
}

/**
Used internally by ParallelBcrypt.Hash()
*/
class ParallelBcryptThread extends Thread {
	final byte[] plainText;
	final int threadIndex;
	final byte[] salt;
	final int cost;

	/*the resulting bcrypt base64 with the salt characters removed*/
	String result;
	
	RuntimeException failReason;	
	
	ParallelBcryptThread(byte[] plainText, int threadIndex, byte[] salt, int cost) {
		this.plainText = plainText;
		this.threadIndex = threadIndex;
		this.salt = salt;
		this.cost = cost;
	}

	public void run() {
		try {
			//Derive a distinct password and salt for this thread to work on:
			byte[] threadByte = new byte[]{(byte)(threadIndex + 1)};
			byte[] threadPassword = Util.hmacSha256(plainText, threadByte);
			
			byte[] threadSalt = new byte[ParallelBcrypt.BCRYPT_SALT_SIZE];
			System.arraycopy(Util.hmacSha256(salt, threadByte), 0, threadSalt, 0, threadSalt.length);

			//Some bcrypt implementations are broken (eg PHP) because they truncate
			// the password at the first null byte!  Therefore I'll pass 64 hex characters.
			//(bcrypt can handle up to 72 bytes)
			String hexPass = Util.hexEncode(threadPassword);

			//ensure lower case
			hexPass = hexPass.toLowerCase();
			if (hexPass.length() != 64) {
				failReason = new IllegalStateException("bad hex encoding");
				return;
			}
		
			String b64 = ParallelBcrypt.standard_bcrypt(hexPass, threadSalt, cost);

			//All bcrypt hashes are 60 characters.
			if (b64.length() != 60)
				throw new IllegalStateException("wrong bcrypt output length");

			//remove the salt and cost prefix (first 29 chars)
			result = b64.substring(29);
		}
		catch (RuntimeException ex) {
			failReason = ex;
		}		
	}

}
